### PR TITLE
fix: MoE aux-loss dtype mismatch under activation checkpointing

### PR DESCRIPTION
**Pull request by pzelasko**
# What does this PR do ?

 ## Summary

Resolves an issue when Automodel is used together with PyTorch Lightning's `bf16-true` precision (`torch.set_default_dtype(torch.bfloat16)` context manager around forward), FSDP2, and activation checkpointing. Tested with Nemotron-3-Nano.

# Changelog

- Fix MoE aux-loss dtype mismatch under activation checkpointing, FSDP2, and torch.bfloat16 default dtype

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by pzelasko** [comment]
I'll test this with a full run but moe aux_loss + bf16 true half-precision + activation checkpointing are broken atm

**Comment by pzelasko** [comment]
@akoumpa this is ready to review and merge - I pinpointed the issue specifically to the combination of activation checkpointing and Lightning's bf16-true https://github.com/Lightning-AI/pytorch-lightning/blob/78bf0214a0ad7571391619dca952c13988c0dc51/src/lightning/pytorch/plugins/precision/half.py#L54-L67

**Comment by pzelasko** [comment]
Validated in SpeechLLM training using Nemotron-3-Nano as LLM backbone (activation checkpointing + PTL's bf16-true + FSDP2 + EP)
image
image

**Review by hemildesai — APPROVED** [review]
